src/main/scala/rise/core/types/Constraints.scala

```diff
@@ -0,0 +1,490 @@
+        case (p: arithexpr.arithmetic.IntDiv, _) => nat.unifyProd(p, b)
+        case (_, p: arithexpr.arithmetic.IntDiv) => nat.unifyProd(p, a)
+        case _ => error(s"cannot unify $a and $b")
+      }
+    }
+
+    def freeOccurences(n: Nat): Map[NatIdentifier, Integer] = {
+      val free_occurrences = mutable
+        .Map[NatIdentifier, Integer]()
+        .withDefault(_ => 0)
+      ArithExpr.visit(n, {
+        case v: NatIdentifier if !v.isExplicit => free_occurrences(v) += 1
+        case _                                 =>
+      })
+      free_occurrences.toMap
+    }
+
+    // collect free variables with only 1 occurrence
+    def potentialPivots(n: Nat): Set[NatIdentifier] = {
+      freeOccurences(n).foldLeft(Set[NatIdentifier]())({
+        case (potential, (v, c)) =>
+          if (c == 1) {
+            potential + v
+          } else {
+            potential
+          }
+      })
+    }
+
+    @scala.annotation.tailrec
+    def pivotSolution(
+                       pivot: NatIdentifier,
+                       n: Nat,
+                       value: Nat
+                     ): Option[Solution] = {
+      n match {
+        case i: NatIdentifier if i == pivot => Some(Solution.subs(pivot, value))
+        case Prod(terms) =>
+          val (p, rest) = terms.partition(t => ArithExpr.contains(t, pivot))
+          if (p.size != 1) {
+            None
+          } else {
+            pivotSolution(pivot, p.head, rest.foldLeft(value)({
+              case (v, r) => v /^ r
+            }))
+          }
+        case Sum(terms) =>
+          val (p, rest) = terms.partition(t => ArithExpr.contains(t, pivot))
+          if (p.size != 1) {
+            None
+          } else {
+            pivotSolution(pivot, p.head, rest.foldLeft(value)({
+              case (v, r) => v - r
+            }))
+          }
+        case Pow(b, Cst(-1)) => pivotSolution(pivot, b, Cst(1) /^ value)
+        case _               => None
+      }
+    }
+
+    def tryPivots(n: Nat, value: Nat)(
+      implicit trace: Seq[Constraint]
+    ): Solution = {
+      potentialPivots(n).foreach(pivotSolution(_, n, value) match {
+        case Some(s) => return s
+        case None    =>
+      })
+      error(s"could not pivot $n = $value")
+    }
+
+    def unifyProd(p: Nat, n: Nat)(implicit trace: Seq[Constraint]): Solution = {
+      // n = p --> 1 = p * (1/n)
+      tryPivots(p /^ n, 1)
+    }
+
+    def unifySum(s: Sum, n: Nat)(implicit trace: Seq[Constraint]): Solution = {
+      // n = s --> 0 = s + (-n)
+      /** TODO(federico) I have no idea how this pivot system works for
+       *  unifying natural numbers. I need to solve a case like
+       *  x + 1 = y + 1. The bruteForceSum function does exactly that:
+       *  see if the pattern matches for some constant, and then unify
+       *  x and y.
+       *
+       *  Upon understanding the pivoting system, this should be cleaned up.
+       */
```

Comment:
Back when I started coding with the dependent pairs I was getting unification problems right there. I went back to check the code snippet, and now it seems the programs no longer take that path - bruteForceSum never seems to be invoked, even though I am using the same tests.

I believe for now I am happy removing the brute forcing and relying on the general strategy - with the caveat that we might want to re-discuss this later in case the problem shows up again